contracts/contracts/validator-registry/middleware/MevCommitMiddleware.sol

```diff
@@ -0,0 +1,704 @@
+    function registerVaults(address[] calldata vaults, uint256[] calldata slashAmounts) external onlyOwner {
+        uint256 vLen = vaults.length;
+        require(vLen == slashAmounts.length, InvalidArrayLengths(vLen, slashAmounts.length));
+        for (uint256 i = 0; i < vLen; ++i) {
+            _registerVault(vaults[i], slashAmounts[i]);
+        }
+    }
+
+    /// @notice Updates the slash amounts for vaults, restricted to contract owner.
+    /// @param vaults Addresses of the vaults to update.
+    /// @param slashAmounts Corresponding slash amounts for each vault.
+    function updateSlashAmounts(address[] calldata vaults, uint256[] calldata slashAmounts) external onlyOwner {
+        uint256 vLen = vaults.length;
+        require(vLen == slashAmounts.length, InvalidArrayLengths(vLen, slashAmounts.length));
+        for (uint256 i = 0; i < vLen; ++i) {
+            _updateSlashAmount(vaults[i], slashAmounts[i]);
+        }
+    }
+
+    /// @notice Requests vault deregistrations, restricted to contract owner.
+    /// @param vaults Addresses of the vaults to request deregistrations for.
+    function requestVaultDeregistrations(address[] calldata vaults) external onlyOwner {
+        uint256 len = vaults.length;
+        for (uint256 i = 0; i < len; ++i) {
+            _requestVaultDeregistration(vaults[i]);
+        }
+    }
+
+    /// @notice Deregisters vaults, restricted to contract owner.
+    /// @param vaults Addresses of the vaults to deregister.
+    function deregisterVaults(address[] calldata vaults) external onlyOwner {
+        uint256 len = vaults.length;
+        for (uint256 i = 0; i < len; ++i) {
+            _deregisterVault(vaults[i]);
+        }
+    }
+
+    /// @notice Registers validators via their BLS public key and vault which will secure them.
+    /// @dev This function is callable by any delegated operator on behalf of a vault.
+    /// @param blsPubkeys BLS public keys of the validators to register.
+    /// @param vaults Addresses of vaults which will secure groups of validators.
+    function registerValidators(bytes[][] calldata blsPubkeys,
+        address[] calldata vaults) external whenNotPaused onlyValidBLSPubKeys(blsPubkeys) {
+        uint256 vaultLen = vaults.length;
+        require(vaultLen == blsPubkeys.length, InvalidArrayLengths(vaultLen, blsPubkeys.length));
+        address operator = msg.sender;
+        _checkOperator(operator);
+        for (uint256 i = 0; i < vaultLen; ++i) {
+            address vault = vaults[i];
+            uint256 keyLen = blsPubkeys[i].length;
+            _checkVault(vault);
+            uint256 potentialSlashableVals = _potentialSlashableVals(vault, operator);
+            require(keyLen <= potentialSlashableVals,
+                ValidatorsNotSlashable(vault, operator, keyLen, potentialSlashableVals));
+            for (uint256 j = 0; j < keyLen; ++j) {
+                _addValRecord(blsPubkeys[i][j], vault, operator);
```

Comment:
we could cache `bytes[] calldata pubkeyArray = blsPubkeys[i]`, to reduce redundant reads

Reply:
see https://github.com/primev/mev-commit/commit/50c9f8531083a278a7a57286892b0095eb15bafc